src/coreclr/jit/helperexpansion.cpp

```diff
@@ -515,7 +515,9 @@ bool Compiler::fgExpandThreadLocalAccessForCallNativeAOT(BasicBlock** pBlock, St
     assert(prevBb != nullptr && block != nullptr);
 
     unsigned finalLclNum         = lvaGrabTemp(true DEBUGARG("Final offset"));
-    lvaTable[finalLclNum].lvType = TYP_I_IMPL;
+    // Note, `tlsRoot` refers to the TLS blob object, which is an unpinned managed object,
+    // thus the type of the local is TYP_REF
+    lvaTable[finalLclNum].lvType = TYP_REF;
```

Comment:
Right. The location that we read is in native TLS buffer, but it contains an Object ref to a managed unpinned object which is the storage for all inlined threadstatics that belong to the current thread. 
(runtime reports the location when GC happens, so the part that native data refers an unpinned instance is ok here)

Once we obtain an obj ref to the storage we need to make sure that the ref is GC-tracked.